### PR TITLE
[CR] Filesystem cleanup

### PR DESCRIPTION
**Pull request by bkentel**
Changes
- moved file_finder and file_wrapper to filesystem.
- made get_files_from_path and get_directories_with free functions.
- split the platform dependent section into separate functions for clarity.
- generalized the BFS search to take a predicate.
- made the ordering of files consistent: lexical breadth first.
- changed the function signatures slightly; very minor tweaks at the call sites.
- removed unused includes.
- removed redundant check in worldfactory.cpp

Remaining
- There is still some code that uses the low level APIs directly in a few places; too big a change to include in this.

### CODE REVIEW

File: src/filesystem.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+    if (path.empty()) {
+        return;
+    }
+
+    dir_ptr root {opendir(path.c_str()), closedir};
+    if (!root) {
+        auto const e = errno;
+        DebugLog(DebugLevel::D_WARNING, DebugClass::D_MAIN) << "Warning: error " << e
+            << " couldn't open file " << path.c_str();
```
Comment (BevapDin) [inline src/filesystem.cpp]: No `c_str` needed here. Maybe include the output of `strerror`?

Comment (bkentel) [inline src/filesystem.cpp]: Yep, missed that when I flipped the msg over to the stream version.
I'll also add strerror output as well.

Thanks for catching those.

**Comment by kevingranade** [comment]
The only thing that concerns me here is that we've had issues with iteration order changes causing json load dependency bugs.  Ideally we wouldn't be subject to that kind of bug at all, but we're not there yet, so we need to be careful.

**Comment by bkentel** [comment]
I didn't notice any problems when I did some simple testing, but are there any tests that can be done to help feel confident about this change? The changes here shouldn't break anything that isn't already broken as both the old and new code iterate breadth-first. The only change is that the absolute ordering is now not undefined (as it was warned to be previously).

**Comment by narc0tiq** [comment]
> The only change is that the absolute ordering is now not undefined (as it was warned to be previously).

And still is. I don't see an update of `data/json/LOADING_ORDER.md` in this pull request.

**Comment by bkentel** [comment]
Good point; I'll take care of it.

**Comment by narc0tiq** [comment]
Thanks much.

**Comment by bkentel** [comment]
Changed the docs to reflect the changes.

**Comment by kevingranade** [comment]
I don't see any pending issues, just took a while to review.

**Comment by bkentel** [comment]
Is this ok to merge? It's blocking me from making changes related to #10806

**Pull request merged**